File: Src/java/cql-engine/src/main/java/org/cqframework/cql/runtime/Concept.java
package org.cqframework.cql.runtime;

import org.cqframework.cql.elm.execution.In;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Bryn on 4/15/2016.
 */
public class Concept {
    private String display;
    public String getDisplay() {
        return display;
    }
    public void setDisplay(String display) {
        this.display = display;
    }
    public Concept withDisplay(String display) {
        setDisplay(display);
        return this;
    }

    private List<Code> codes = new ArrayList<>();
    public Iterable<Code> getCodes() {
        return codes;
    }
    public void setCodes(Iterable<Code> codes) {
        this.codes.clear();
        for (Code code : codes) {
            this.codes.add(code);
        }
    }
    public Concept withCodes(Iterable<Code> codes) {
        setCodes(codes);
        return this;
    }
    public Concept withCode(Code code) {
        codes.add(code);
        return this;
    }

    @Override
    public boolean equals(Object other) {
        // Concept equality is defined as
        if (other instanceof Concept) {
            Concept otherConcept = (Concept)other;
            for (Code code : this.getCodes()) {
                if (In.in(code, otherConcept.getCodes())) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/runtime/Concept.kt
package org.opencds.cqf.cql.engine.runtime

import kotlin.js.ExperimentalJsExport
import org.cqframework.cql.shared.JsOnlyExport

@OptIn(ExperimentalJsExport::class)
@JsOnlyExport
class Concept : StructuredValue(), NamedTypeValue {
    override val type = conceptTypeName

    override val elements: MutableMap<kotlin.String, Value?>
        get() = mutableMapOf("codes" to codes?.toCqlList())

    var display: kotlin.String? = null

    fun withDisplay(display: kotlin.String?): Concept {
        this.display = display
        return this
    }

    var codes: MutableList<Code?>? = mutableListOf<Code?>()
        set(codes) {
            field!!.clear()
            if (codes != null) {
                for (code in codes) {
                    field!!.add(code)
                }
            }
        }

    @Suppress("NON_EXPORTABLE_TYPE")
    fun withCodes(codes: Iterable<Code?>?): Concept {
        this.codes = codes?.toMutableList() ?: mutableListOf()
        return this
    }

    fun withCode(code: Code?): Concept {
        codes!!.add(code!!)
        return this
    }

    override fun toString(): kotlin.String {
        return toPrettyString("Concept")
    }
}
